spark/v3.4/spark-extensions/src/main/scala/org/apache/iceberg/spark/extensions/IcebergMergeInto.scala

```diff
@@ -0,0 +1,399 @@
+import org.apache.spark.sql.catalyst.plans.logical.InsertStarAction
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.catalyst.plans.logical.MergeAction
+import org.apache.spark.sql.catalyst.plans.logical.MergeIntoContext
+import org.apache.spark.sql.catalyst.plans.logical.UnresolvedMergeIntoIcebergTable
+import org.apache.spark.sql.catalyst.plans.logical.UpdateAction
+import org.apache.spark.sql.catalyst.plans.logical.UpdateStarAction
+import org.apache.spark.sql.functions.expr
+import scala.collection.JavaConverters
+
+
+object IcebergMergeInto {
+  /**
+   * Initialize an [[IcebergMergeIntoBuilder]].
+   *
+   * @param table : Target table name of the merge action
+   * @return [[IcebergMergeIntoBuilder]]
+   */
+  def apply(table: String): IcebergMergeIntoBuilder =
+    new IcebergMergeIntoBuilder(table, None, None, Seq.empty[MergeAction], Seq.empty[MergeAction])
+
+  /**
+   * Builder to specify an IcebergMergeInto action.
+   *
+   * It could be possible to provide any number of `whenMatched`
+   * and `whenNotMatched` actions.
+   *
+   *
+   * Scala Examples:
+   * {{{
+   *   val ds = ...
+   *   IcebergMergeInto("icebergTable")
+   *      .using(ds.as("source")
+   *      .when("source.id = icebergTable.id")
+   *      .whenMatched("source.op = U")
+   *      .updateAll()
+   *      .whenMatched("source.op = D)
+   *      .delete()
+   *      .whenNotMatched()
+   *      .insertAll()
+   * }}}
+   *
+   * JavaExamples:
+   * {{{
+   *   Dataset<Row> ds = ...
+   *   IcebergMergeInto.apply("icebergTable")
+   *      .using(ds.as("source")
+   *      .when("source.id = icebergTable.id")
+   *      .whenMatched("source.op = U")
+   *      .updateAll()
+   *      .whenMatched("source.op = D)
+   *      .delete()
+   *      .whenNotMatched()
+   *      .insertAll()
+   * }}}
+   *
+   */
+  class IcebergMergeIntoBuilder(
+                                 private val targetTable: String,
+                                 private val source: Option[Dataset[Row]],
+                                 private val whenCondition: Option[Expression],
+                                 private val whenMatchedActions: Seq[MergeAction],
+                                 private val whenNotMatchedActions: Seq[MergeAction]) {
+
+    /**
+     * Set the source dataset used during merge actions.
+     *
+     * @param source  : Dataset[Row]
+     * @return [[IcebergMergeIntoBuilder]]
+     */
+    def using(source: Dataset[Row]): IcebergMergeIntoBuilder =
+      new IcebergMergeIntoBuilder(
+        targetTable,
+        Some(source),
+        whenCondition,
+        whenMatchedActions,
+        whenNotMatchedActions
+      )
+
+    /**
+     * Set the `when` condition of the merge action from a [[Column]]
+     *
+     * @param condition : [[Column]] expression
+     * @return [[IcebergMergeIntoBuilder]]
+     */
+    def when(condition: Column): IcebergMergeIntoBuilder =
```

Comment:
The `when` seems a little confused with `whenMatched` / `whenNotMatched`. What about `on`?

Reply:
@ConeyLiu I agree. 
I use when to have similarity with Delta MergeApi. But I completely agree that on could be better then when.

Reply:
I would keep this as "on" because fundamentally this is setting shuffle join conditions (join on) while the other functions are only exercised at execution time.

Reply:
Hi @ConeyLiu and @RussellSpitzer this should be fixed now